### DISCUSSION TITLE
start() vs start_soon()?

### DISCUSSION THREAD

**Discussion by jorgesca** [cocotb/cocotb#4215] [Q&A] [answered]
I am new to cocotb, and new to coroutines in general. The docs state:

> The async function [start()](https://docs.cocotb.org/en/stable/library_reference.html#cocotb.start) schedules the new coroutine to be executed concurrently, then yields control to allow the new task (and any other pending tasks) to run, before resuming the calling task.

> [start_soon()](https://docs.cocotb.org/en/stable/library_reference.html#cocotb.start_soon) schedules the new coroutine for future execution, after the calling task yields control.

But I can't grasp the difference, maybe because the coroutine concept is new to me. I've read in an [issue](https://github.com/cocotb/cocotb/pull/2713#issuecomment-923338510) that "`start` is just `start_soon` + `await NullTrigger`", but I don't get what `NullTrigger` is either, so....
I'd appreciate some hints to educate me on this. Thanks

**Comment by marlonjames** [comment, marked as answer]
`start_soon()` will queue up the new `Task` to start running after the next trigger you `await`, so whatever code is in your task after `start_soon()` but before you await a trigger gets to run *before* the new task starts.

If you want the new task to start and run before you continue running the current task, use `await start()`. Just remember that the new task gets to run until it awaits a trigger.

**Reply by jorgesca** [reply]
Thanks!
